### PR TITLE
Fix OTLP testing flake: signal connection from mock collector

### LINKED ISSUE #1527: Flaky test: TestNewExporter_collectorConnectionDiesThenReconnects
Reporting to be failing about %5 of the time on Go 1.15 but not seen on 1.14.

Version tested: https://github.com/open-telemetry/opentelemetry-go/commit/e50a1c8cecb8c584bdfada6cb21fe0b173a22646

```
% docker run -v $(pwd)/opentelemetry-go:/otsrc --rm -w /otsrc/exporters/otlp/otlpgrpc golang:1.15 go test -run=TestNewExporter_collectorConnectionDiesThenReconnects
go: downloading google.golang.org/grpc v1.35.0
go: downloading github.com/gogo/protobuf v1.3.2
punya@obsidian ~ % docker run -v $(pwd)/opentelemetry-go:/otsrc --rm -w /otsrc/exporters/otlp/otlpgrpc golang:1.15 go test -run=TestNewExporter_collectorConnectionDiesThenReconnects -count 100 -v
go: downloading google.golang.org/grpc v1.35.0
go: downloading github.com/stretchr/testify v1.7.0
go: downloading github.com/gogo/protobuf v1.3.2
go: downloading gopkg.in/yaml.v3 v3.0.0-20200313102051-9f266ea9e77c
go: downloading github.com/pmezard/go-difflib v1.0.0
go: downloading github.com/davecgh/go-spew v1.1.0
go: downloading google.golang.org/genproto v0.0.0-20200526211855-cb27e3aa2013
go: downloading github.com/golang/protobuf v1.4.2
go: downloading golang.org/x/net v0.0.0-20201021035429-f5854403a974
go: downloading golang.org/x/sys v0.0.0-20200930185726-fdedc70b468f
go: downloading google.golang.org/protobuf v1.25.0
go: downloading golang.org/x/text v0.3.3
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
    otlp_integration_test.go:182: 
                Error Trace:    otlp_integration_test.go:182
                Error:          Received unexpected error:
                                rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:43939: connect: connection refused"
                Test:           TestNewExporter_collectorConnectionDiesThenReconnects
--- FAIL: TestNewExporter_collectorConnectionDiesThenReconnects (0.73s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
    otlp_integration_test.go:182: 
                Error Trace:    otlp_integration_test.go:182
                Error:          Received unexpected error:
                                rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:36879: connect: connection refused"
                Test:           TestNewExporter_collectorConnectionDiesThenReconnects
--- FAIL: TestNewExporter_collectorConnectionDiesThenReconnects (0.49s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
    otlp_integration_test.go:182: 
                Error Trace:    otlp_integration_test.go:182
                Error:          Received unexpected error:
                                rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:36283: connect: connection refused"
                Test:           TestNewExporter_collectorConnectionDiesThenReconnects
--- FAIL: TestNewExporter_collectorConnectionDiesThenReconnects (0.24s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
    otlp_integration_test.go:182: 
                Error Trace:    otlp_integration_test.go:182
                Error:          Received unexpected error:
                                rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:40819: connect: connection refused"
                Test:           TestNewExporter_collectorConnectionDiesThenReconnects
--- FAIL: TestNewExporter_collectorConnectionDiesThenReconnects (0.24s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
    otlp_integration_test.go:182: 
                Error Trace:    otlp_integration_test.go:182
                Error:          Received unexpected error:
                                rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:37631: connect: connection refused"
                Test:           TestNewExporter_collectorConnectionDiesThenReconnects
--- FAIL: TestNewExporter_collectorConnectionDiesThenReconnects (0.49s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.89s)
=== RUN   TestNewExporter_collectorConnectionDiesThenReconnects
--- PASS: TestNewExporter_collectorConnectionDiesThenReconnects (0.90s)
FAIL
exit status 1
FAIL    go.opentelemetry.io/otel/exporters/otlp/otlpgrpc        87.372s
```

### PR DESCRIPTION
**Pull request by MrAlias**
Add a TCP Listener implementation to the mock collector that signals when it receives a connection. Use this signaling in testing to wait for re connections to deterministically happen instead of waiting an arbitrary time and hoping it happens.

Validated this fix [here](https://github.com/open-telemetry/opentelemetry-go/pull/1815).

Resolve #1527

### CODE REVIEW

File: exporters/otlp/otlpgrpc/mock_collector_test.go [hunk 1]
```diff
@@ -189,19 +205,62 @@ func runMockCollectorAtEndpoint(t *testing.T, endpoint string) *mockCollector {
...
+	return &listener{
+		wrapped: wrapped,
+		C:       make(chan struct{}, 1),
+		closed:  make(chan struct{}),
 	}
+}
 
-	_, collectorPortStr, _ := net.SplitHostPort(ln.Addr().String())
+func (l *listener) Accept() (net.Conn, error) {
+	conn, err := l.wrapped.Accept()
```
Comment (MadVikingGod) [inline exporters/otlp/otlpgrpc/mock_collector_test.go:238]: This should return an error when wrapped is closed.  Is it possible to use that as the signal that we have called closed?

File: exporters/otlp/otlpgrpc/mock_collector_test.go [hunk 2]
```diff
@@ -174,6 +177,19 @@ func (mc *mockCollector) GetMetrics() []*metricpb.Metric {
...
+// with the mockCollector before returning.
+func (mc *mockCollector) WaitForConn() {
+	for {
+		select {
+		case <-mc.ln.C:
+			return
+		default:
+			runtime.Gosched()
+		}
+	}
```
Comment (MadVikingGod) [inline exporters/otlp/otlpgrpc/mock_collector_test.go:190]: Nitpick.  This logic seems more at home in the listener, and should just be a proxy call here.

**Review by MadVikingGod — COMMENTED** [review]
Looks good, I will be happy to rid ourselves of a flaky test.

**Review by Aneurysm9 — APPROVED** [review]

**Review by XSAM — APPROVED** [review]
👍

**Pull request merged**